### ISSUE TITLE
Cannot attach network to stopped container with --net none

### ISSUE THREAD

**Issue by sheerun** [docker/docker#21132]
**Output of `docker version`:**

```
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.6
 Git commit:   c3959b1
 Built:        Tue Feb 23 21:11:36 UTC 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 21:49:11 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 11
 Running: 1
 Paused: 0
 Stopped: 10
Images: 6
Server Version: 1.10.3
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 78
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: weavemesh bridge null host
Kernel Version: 4.1.19-boot2docker
Operating System: Boot2Docker 1.10.3 (TCL 6.4.1); master : 625117e - Thu Mar 10 22:09:02 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.523 GiB
Name: default
ID: DNVE:MHQ6:R3KE:PWMV:AC7P:T27J:NSZS:NDX5:X5FC:6GFA:BWLA:2WVU
Debug mode (server): true
 File Descriptors: 20
 Goroutines: 33
 System Time: 2016-03-11T19:46:17.855346239Z
 EventsListeners: 1
 Init SHA1:
 Init Path: /usr/local/bin/docker
 Docker Root Dir: /mnt/sda1/var/lib/docker
Username: sheerun
Registry: https://index.docker.io/v1/
Labels:
 provider=parallels
```

**Steps to reproduce the issue:**

```
docker network create --driver overlay --subnet 10.40.1.0/16 --internal net1
docker network create --driver overlay --subnet 10.50.1.0/16 --internal net2
docker create --name busybox -ti --net none busybox ifconfig
docker network connect --ip 10.40.0.5 net1 busybox
docker network connect --ip 10.50.0.5 net2 busybox
docker start -ai busybox
```

**Describe the results you received:**

> Error response from daemon: Container cannot be connected to multiple networks with one of the networks in private (none) mode

**Describe the results you expected:**

Docker is connected is started with 2 networks connected.

**Additional information you deem important (e.g. issue happens only occasionally):**

The only method that works is following:

```
docker create --name busybox -ti --net net1 --ip 10.40.1.0 busybox ifconfig
docker network connect --ip 10.40.1.0 net2 busybox
docker start -ai busybox
```

I find this method of attaching multiple networks inconsistent... As you always need to list first network with `--net` and `--ip`, instead just creating container and attaching multiple networks the same way..

**Comment by WeiZhang555** [comment]
I think you can have another choice, run
```
$ docker network disconnect none busybox
```
Before you try to attach this container to two overlay networks. 

Guess this is by design, container in network `none` indicates it can't join any network.
Is this acceptable for you?

**Comment by sheerun** [comment]
I guess it's good enough for me. I leave you the decision to either close this issue or improve the API.

**Comment by WeiZhang555** [comment]
@sheerun I'm not a Docker maintainer so can't close this issue.

I'm just offering some suggestions, so if you have doubt about the answer, you can still ping other maintainers for better solution :smile:

**Comment by thaJeztah** [comment]
Hm, although this may have a technical reason, from a UX perspective I think this is surprising; creating a container without network (`--net none`), and attaching it to a network later on sounds like something I should be able to do.

ping @mavenugo is there a reason we don't (or can't) support this?

**Comment by cirocosta** [comment]
This is labelled `version/1.10`, is there any intention on reviewing this? I think it makes sense to have it like @thaJeztah said, even though it also makes sense to me having `net=none` as a way os explicitly blocking it from connecting to networks later ...

Anyway, it's not possible right now to create a container that is attached to 0 networks and then attach it to a user-defined network, right?

**Comment by thaJeztah** [comment]
@cirocosta the `version/1.10` label is automatically added based on the version that an issue was first reported with. I'll try to get clear why this is currently not supported

**Comment by aboch** [comment]
I agree with @WeiZhang555 because of the special meaning of the network mode `none`.
In my opinion a container can be connected to multiple networks, yes, but it cannot at the same time be in `no network mode` and also be connected to networks. Those are two incompatible states.

**Comment by thaJeztah** [comment]
@aboch from a "user perspective", `--network=none` feels as "don't connect it to a network", not as _"connect it to the network "no networking""_. Doing `docker network connect` then would be "there's no networks connected yet, so now connect it to the network I specified"

**Comment by aboch** [comment]
@thaJeztah 
I completely agree with you point about user perspective, if we are ok on modifying the API semantic.
It may also be people would not miss current behavior.
We probably need more input from others before deciding on this.

**Comment by mryan43** [comment]
I have a case in my CI environments where a build container managed by kubernetes needs to be connected to a docker network to have DNS and connectivity for some e2e tests environments.

It works fine by doing a "docker network connect" between the "build" pod's infra container and the custom docker network as long as we don't enable a CNI plugin in kubernetes. When enabling the calico CNI plugin, the pod's infra container gets created with NetworkMode="none" and the "docker network connect" fails with this error : Container cannot be connected to multiple networks with one of the networks in private (none) mode

I feel like this restriction is arbitrary and there is no technical blocker for docker to create the new interface in the container and attach it to the custom network's bridge.

This is currently preventing me from enabling support for NetworkPolicies in my kubernetes cluster T_T by switching the network fabric from flannel to calico

**Comment by chrisnew** [comment]
I agree with @mryan43 

I also came across this message today and that behavior renders SaltStack’s container network management partially unusuable.
